### ISSUE TITLE
RuleTester feature: `setup` for test cases

### ISSUE THREAD

**Issue by RobinTail** [eslint/eslint#18770]
### ESLint version

v9.9.0

### What problem do you want to solve?

I'm developing a plugin that must be aware of the user's `package.json`.
The plugin reads that file using `readFileSync`.
In order to test that plugin I need to run the tests cases having not only `options` but also mocking that `package.json` for different scenarios.

I find it excessive and overcomplicating the suggested approach on fixtures in form of actual files in multiple directories that have to be maintained and being detached from the test cases. I have a better solution, that I believe will improve the developer experience.

### What do you think is the correct solution?

Image you can do this:

```ts
const readerMock = vi.fn();

vi.mock("node:fs", () => ({
  readFileSync: readerMock,
}));

const ruleTester = new RuleTester();

ruleTester.run("rule name", rule, {
  valid: [
    {
      code: `some code`,
      options: { some: "options" },
      // the proposed property:
      setup: () => { 
        readerMock.mockReturnValueOnce(
          JSON.stringify( { dependencies: {eslint: "^9" } } )
        );
      }
    },
    // ... more others with different setup and options
  ],
  invalid: [ /* similar */ ]
});
```

Thus, I could keep the mocked environment right within the test case.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

I recently [made a similar suggestion](https://github.com/typescript-eslint/typescript-eslint/issues/9770) to `typescript-eslint`, but I was advised to refer to `eslint` first.

**Comment by nzakas** [comment]
Thanks for the suggestion. I'm having trouble understanding what `setup` is supposed to do or enable. Is there a reason you can't use `beforeAll()` or `beforeEach()` to achieve this? If not, can you please explain more about what it is you're trying to do and what existing approaches you've tried?

**Comment by RobinTail** [comment]
@nzakas , this can not be done using `beforeEach`, because `beforeEach` is something similar you do before every test.
`setup` is as unique as `options`.
Me, personally, could use `setup` to mock `package.json` for every individual test case instead of making dozens of physical files.

**Comment by nzakas** [comment]
Thanks. I'm just trying to understand better how this applies to more than just your specific case.

It sounds like you're saying that you'd like a hook for each test case that runs before the actual linting, is that correct?

Would it also then stand to reason that one might want a hook to run after each test case to clean up whatever was done before?

FWIW, for changes like this, we do require the request to go through the [RFC process](https://github.com/eslint/rfcs) to make sure we are considering all of the implications.

**Comment by RobinTail** [comment]
> how this applies to more than just your specific case.

it applies at least to `eslint-import-plugin`, for example, that reads `package.json` for its `no-extranerous-dependencies` rule, but has to maintain fixture-files in its tests, @nzakas , check out these two links:

- reading: https://github.com/import-js/eslint-plugin-import/blob/09476d7dac1ab36668283f9626f85e2223652b37/src/rules/no-extraneous-dependencies.js#L23
- fixture files: https://github.com/import-js/eslint-plugin-import/blob/09476d7dac1ab36668283f9626f85e2223652b37/tests/src/rules/no-extraneous-dependencies.js#L21-L29

**And what I'm proposing is supposed to make the developer experience better.**

> It sounds like you're saying that you'd like a hook for each test case that runs before the actual linting, is that correct?

Yes. I made a PR 38 lines long, that does exactly what I mean.

> Would it also then stand to reason that one might want a hook to run after each test case to clean up whatever was done before?

Well, I didn't find it necessary, because all the testing framework I use have a capability to mock result/implementation once, therefore I didn't encounter a need for `teardown`. However, as you said "one might want" and perhaps that it could make sense having a symmetry in that matter.

> we do require the request to go through the [RFC process](https://github.com/eslint/rfcs)

OK, I will make one then, @nzakas

**Comment by nzakas** [comment]
Thanks, that's some helpful context. I appreciate that you want to make the developer experience better, we just need to fully understand what it is you're proposing and to make sure it's helpful to not just you but to the developer ecosystem as a whole.

Looking forward to the RFC.

**Comment by RobinTail** [comment]
Did my best in https://github.com/eslint/rfcs/pull/122 , @nzakas

**Comment by RobinTail** [comment]
It's in progress there https://github.com/eslint/rfcs/pull/122
